File: app/src/main/java/com/suzei/minote/data/DataSourceImpl.java
package com.suzei.minote.data;

import com.suzei.minote.data.dao.NotesDao;
import com.suzei.minote.data.entity.Notes;
import com.suzei.minote.utils.executors.AppExecutor;

import java.util.List;

public class DataSourceImpl implements DataSource {

    private static DataSourceImpl sInstance = null;

    private NotesDao notesDao;

    private AppExecutor appExecutor;

    public DataSourceImpl(AppExecutor appExecutor, NotesDao notesDao) {
        this.appExecutor = appExecutor;
        this.notesDao = notesDao;
    }

    public static DataSourceImpl getInstance(AppExecutor appExecutor, NotesDao notesDao) {
        if (sInstance == null) {
            sInstance = new DataSourceImpl(appExecutor, notesDao);
        }

        return sInstance;
    }

    @Override
    public void saveNote(Notes note) {
        Runnable runnable = () -> notesDao.saveNote(note);
        appExecutor.getDiskIO().execute(runnable);
    }

    @Override
    public void getNote(int itemId, NoteListener listener) {
        Runnable runnable = () -> {
            Notes note = notesDao.findNoteById(itemId);

            appExecutor.getMainThread().execute(() -> {
                if (note == null) {
                    listener.onDataUnavailable();
                }
                else {
                    listener.onDataAvailable(note);
                }
            });
        };

        appExecutor.getDiskIO().execute(runnable);
    }

    @Override
    public void getListOfNotes(ListNoteListener listener) {
        Runnable runnable = () -> {

            List<Notes> listOfNotes = notesDao.findAllNotes();

            appExecutor.getMainThread().execute(() -> {

                if (listOfNotes.isEmpty()) {
                    listener.onDataUnavailable();
                } else {
                    listener.onDataAvailable(listOfNotes);
                }

            });

        };

        appExecutor.getDiskIO().execute(runnable);
    }

    @Override
    public void deleteNote(Notes note) {
        Runnable runnable = () -> notesDao.deleteNote(note);
        appExecutor.getDiskIO().execute(runnable);
    }

}

File: app/src/main/java/com/suzei/minote/data/DataSourceImpl.kt
package com.suzei.minote.data

import com.suzei.minote.data.dao.NotesDao
import com.suzei.minote.data.entity.Notes
import com.suzei.minote.utils.executors.AppExecutor

class DataSourceImpl private constructor(
        private val appExecutor: AppExecutor,
        private val notesDao: NotesDao) : DataSource {

    companion object {

        fun getInstance(appExecutor: AppExecutor, notesDao: NotesDao): DataSourceImpl {
            return DataSourceImpl(appExecutor, notesDao)
        }

    }

    override fun updateNote(note: Notes) {
        val runnable = { notesDao.updateNote(note) }
        appExecutor.diskIO.execute(runnable)
    }

    override fun saveNote(note: Notes) {
        val runnable = { notesDao.saveNote(note) }
        appExecutor.diskIO.execute(runnable)
    }

    override fun deleteNote(note: Notes) {
        val runnable = { notesDao.deleteNote(note) }
        appExecutor.diskIO.execute(runnable)
    }

    override fun getNote(itemId: Int, listener: DataSource.NoteListener) {
        val runnable = {
            val note = notesDao.findNoteById(itemId)

            appExecutor.mainThread.execute {
                listener.onDataAvailable(note)
            }
        }

        appExecutor.diskIO.execute(runnable)
    }

    override fun getListOfNotes(listener: DataSource.ListNoteListener) {
        val runnable = {

            val listOfNotes = notesDao.findAllNotes()

            appExecutor.mainThread.execute {

                if (listOfNotes.isEmpty()) {
                    listener.onDataUnavailable()
                } else {
                    listener.onDataAvailable(listOfNotes.toMutableList())
                }

            }

        }

        appExecutor.diskIO.execute(runnable)
    }

}
